### PR TITLE
Include lambdas in grammars for functions

### PR DESCRIPTION
**Pull request by ajreynol**
Allows us to solve some corner case HO benchmarks using `--sygus-inst`, one is added in this PR.

Also fixes a bug in ground term purification that can arise in HO.

### CODE REVIEW

File: test/regress/cli/regress1/quantifiers/ho-seu-sygus-inst.smt2
Comment (abdoo8080) [inline test/regress/cli/regress1/quantifiers/ho-seu-sygus-inst.smt2:1]: Is this file tested?

File: src/theory/quantifiers/ematching/trigger.cpp [hunk 1]
```diff
@@ -199,7 +199,7 @@ Node Trigger::ensureGroundTermPreprocessed(Valuation& val,
     it = visited.find(cur);
     if (it == visited.end())
     {
-      if (cur.getNumChildren() == 0)
+      if (cur.getNumChildren() == 0 || cur.getKind()==Kind::BOUND_VAR_LIST)
```
Comment (abdoo8080) [inline src/theory/quantifiers/ematching/trigger.cpp:202]: Format...

**Review by abdoo8080 — APPROVED** [review]
LGTM! Just address the comments.
